### ISSUE TITLE
Adding distance/cost to the connection rather than in point AStar.

### ISSUE THREAD

**Issue by pegasusearl** [godotengine/godot-proposals#334]
## **Describe the project you are working on:**
Restaurant simulator where staff and customer will need pathfinding to walk from door, kitchen, seat and table. The game is about **effective restaurant layout** so abit more accurate pathfinding will be useful.

## **Describe the problem or limitation you are having in your project:**
You can't set `connection_cost` (from point to point) manually. A point can have multiple **connections** that will have different **distance** so we can't just use `weight_scale` for it. `weight_scale` is setting cost at the **point**, not in the **connection**. In current astar, it's impossible to set `connection_cost`.
![astarcost](https://user-images.githubusercontent.com/23167087/71421286-7eb14a80-26ac-11ea-97bd-e92fa48a8ece.png)
Usually I just ignore `weight_scale` or distance but for this project, I wish I could make better one.

## **Describe how this feature / enhancement will help you overcome this problem or limitation:**
If we can manually set `connection_cost`. We won't need hacky workaround.
![astarcostz](https://user-images.githubusercontent.com/23167087/71421322-b9b37e00-26ac-11ea-96bf-a1194e72e1a9.png)

### Current

We can set weight_scale, but we can't manually set distance between points. Distance determined by weight_scale.
![without](https://user-images.githubusercontent.com/23167087/71421365-fda68300-26ac-11ea-9137-da3e213bf30a.png)

### With customizable connection_cost

If we can manually set the distance between point, we can do this:
![with](https://user-images.githubusercontent.com/23167087/71421379-09924500-26ad-11ea-8db6-ded090cc72dc.png)
see the distance.

Circle with A B  C is points. and the arrow is the connections.

## **Describe implementation detail for your proposal (in code), if possible:**
Unfortunately, I don't know how to implement it in the engine. But this is how godot user can use it:
```
add_point ( int id, Vector3 position, float weight_scale=1.0 ) # add point stay the same just incase some user actually used weight_scale,
connect_points ( int id, int to_id, bool bidirectional=true, <float/int> cost=1 ) # just adding cost
```

## **If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No. It can be workarounded but definitely not with just few lines of code.
What I can think of is adding an "inbetween node" for connections that had extra distance. Then everytime we call `get_point_path()` we will need to iterate the result and check if there is "inbetween node" then remove them. Of course that also means we will have list of which one is regular point and which one is inbetween.

## **Is there a reason why this should be core and not an add-on in the asset library?:**
To my **subjective** opinion, I don't undrestand why we have `weight_scale` instead  of custom distance. Custom distance easily workaround weight_scale but weight_scale cannot fix different distances in the from or to the same points.
`weight_scale` had far fewer usecase than custom distance. Example are the pictures I shown above, diagonal movement and point to point movement. (that point to point movement can be used for world_map pathfinding. World map is set of areas in which distance between areas to areas is different)
I can't find any *general use* of `weight_scale` at all.

Thank you for reading this far. It's alot of text.
I apologize if there is something weird or wrong with my request. But if there is any, I would like to hear it.

**Comment by logzero** [comment]
Yeah, having a point weight seems to be a bit of a weird choice, a memory usage optimization?

From a quick look at the code you'd have to move weight_scale from Point to Segment and patch up all the functions using it.

Or to remain backwards compatible, keep Point weight_scale and have additional Segment weight_scale.

**Comment by smix8** [comment]
Both should be available. Point cost / weight as well as connection cost / weight and the option to replace the geometry distance calculation between points with a custom distance value.

While all this is possible if you extend the AStar class and override the `_compute_cost()` and `_estimate_cost()` functions this is too much asked for such a commonly required navigation feature especially for beginners.

**Comment by paul173205** [comment]
Hi @Calinou. Thanks for noticing my post!

I don't see any reason to include the phrase "rather than in point". I don't think it is an "either/or". In my head this new optional argument "weight_scale", on a connection, just runs fully in parallel with the existing weight_scale for points. Either (or both) can be used independently (or together). So I believe a better title is "Allow AStar (2D and 3D) to have weight_scale on connections in addition to the existing weight_scale on points."

In my mind connect_points goes from looking like:
void connect_points(id: int, to_id: int, bidirectional: bool = true)

To looking like:
void connect_points(id: int, to_id: int, bidirectional: bool = true, weight_scale: float = 1.0)

To fully parallel the current use of weight_scale on the points, we'd also need to add:
get_connection_weight_scale(id: int, to_id: int) const
set_connection_weight_scale(id: int, to_id: int, weight_scale: float)

And then where the actual weight is calculated for a segment traversal, this additional weight_scale gets multiplied in as well.

Thanks.
